library/std/src/sys/unix/thread_local_dtor.rs

```diff
@@ -11,28 +11,47 @@
 // Note, however, that we run on lots older linuxes, as well as cross
 // compiling from a newer linux to an older linux, so we also have a
 // fallback implementation to use as well.
+#[allow(unexpected_cfgs)]
 #[cfg(any(target_os = "linux", target_os = "fuchsia", target_os = "redox"))]
 pub unsafe fn register_dtor(t: *mut u8, dtor: unsafe extern "C" fn(*mut u8)) {
     use crate::mem;
     use crate::sys_common::thread_local_dtor::register_dtor_fallback;
 
+    /// This is necessary because the __cxa_thread_atexit_impl implementation
+    /// std links to by default may be a C or C++ implementation that was not
+    /// compiled using the Clang integer normalization option.
+    #[cfg(not(sanitizer_cfi_normalize_integers))]
+    #[cfi_encoding = "i"]
+    #[repr(transparent)]
+    pub struct c_int(pub libc::c_int);
+
     extern "C" {
         #[linkage = "extern_weak"]
         static __dso_handle: *mut u8;
         #[linkage = "extern_weak"]
-        static __cxa_thread_atexit_impl: *const libc::c_void;
+        fn __cxa_thread_atexit_impl(
+            dtor: unsafe extern "C" fn(*mut libc::c_void),
+            arg: *mut libc::c_void,
+            dso_handle: *mut libc::c_void,
+        ) -> c_int;
+    }
+
+    // __cxa_thread_atexit_impl may be null because of the extern_weak linkage
+    fn is_null(f: *mut libc::c_void) -> bool {
+        f.is_null()
```

Comment:
I don't believe this is guaranteed to work.

Reply:
Humm... do you happen to know why it's not guaranteed to work? (I'm trying to find more information about it.)

Reply:
The pattern I recommended was enabled after this comment: https://github.com/rust-lang/rust/issues/29603#issuecomment-1227764764

The reason why it isn't guaranteed to work is because this is the classic case of asking an optimizing compiler if a value is null when it is already guaranteed to be non-null, and that you guaranteed by declaring it to be a non-null type. Even if Rust/LLVM doesn't optimize on this case, I can't guarantee we won't tomorrow, basically.

`#[linkage]` really should be one of the things that trigger the `deny(unsafe_code)` lint, as with the rest here:
- https://github.com/rust-lang/rust/issues/82499
- https://github.com/rust-lang/rust/issues/103078

Or the declaration you just did should be banned (or both).

Reply:
A function item is guaranteed to be non null, but in this case it was used with a weak linkage, which may be null. So, the compiler shouldn't have allowed me to make this declaration using weak linkage, right? If the function item ends up being null, isn't this check going to catch it?

It seems there is also another problem that by using the pattern you suggested the compiler also omits the weakly function definition from the LLVM IR (and its associated metadata).

Reply:
>  So, the compiler shouldn't have allowed me to make this declaration using weak linkage, right? 

Correct, there is at least one compiler bug in play.

> If the function item ends up being null, isn't this check going to catch it?

There is no guarantee. It depends on whether the optimizer notices, because the optimizer is allowed to remove the check.

> It seems there is also another problem that by using the pattern you suggested the compiler also omits the weakly function definition from the LLVM IR (and its associated metadata).

Ah, metadata going missing is why you cannot use it?

And it does not work to use the previous `static` pattern but transmuting the pointer to the matching `unsafe extern "C" fn(unsafe extern "C" fn(*mut c_void), *mut c_void, *mut c_void)` pattern?